openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -0,0 +1,104 @@
+
+
+def upload_to_path(bs_model, filename):  # pylint:disable=W0613
+    """
+    Returns path name to use for the given
+    BlockStructureModel instance.
+    """
+    return unicode(bs_model.data_usage_key)
+
+
+class BlockStructureModel(TimeStampedModel):
+    """
+    Model for storing Block Structure information.
+    """
+    class Meta(object):
+        db_table = 'block_structure'
+
+    data_usage_key = UsageKeyField(
+        u'Identifier of the data being collected.',
+        blank=False,
+        max_length=255,
+        db_index=True,
+    )
+    data_version = models.CharField(
+        u'Version of the data at the time of collection.',
+        blank=True,
+        max_length=255,
+    )
+    data_edit_timestamp = models.DateTimeField(
+        u'Edit timestamp of the data at the time of collection.',
+        blank=True,
+        null=True,
+    )
+
+    transformers_schema_version = models.CharField(
+        u'Representation of the schema version of the transformers used during collection.',
+        blank=False,
+        max_length=255,
+    )
+    block_structure_schema_version = models.CharField(
+        u'Version of the block structure schema at the time of collection.',
+        blank=False,
+        max_length=255,
+    )
+
+    data = models.FileField(upload_to=upload_to_path)
+
+    def get_serialized_data(self):
+        """
+        Returns the collected data for this instance.
+        """
+        return self.data.read()
+
+    @classmethod
+    def get(cls, data_usage_key):
+        """
+        Returns the entry associated with the given data_usage_key.
+        Raises:
+             BlockStructureNotFound if an entry for data_usage_key is not found.
+        """
+        try:
+            return cls.objects.get(data_usage_key=data_usage_key)
+        except cls.DoesNotExist:
+            raise BlockStructureNotFound(data_usage_key)
+
+    @classmethod
+    def update_or_create(cls, serialized_data, **kwargs):
+        """
+        Updates or creates the BlockStructureModel entry
+        for the given data_usage_key in the kwargs,
+        uploading serialized_data as the content data.
+        """
+        bs_model, _ = cls.objects.update_or_create(**kwargs)
+        bs_model.data.save('', ContentFile(serialized_data))
+        return bs_model
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return u', '.join(
+            u'{}: {}'.format(field_name, unicode(getattr(self, field_name)))
+            for field_name in
+            [
+                u'data_usage_key',
+                u'data_version',
+                u'data_edit_timestamp',
+                u'transformers_schema_version',
+                u'block_structure_schema_version',
+            ]
+        )
+
+    # TODO override delete, to delete underyling file object also?
+    # TODO need to prune old data when new course version is created
```

Comment:
It seems like `update_or_create` is the main method by which a `new course version is created` - could we add some special logic in there to try and update* in cases where we match an existing usage key, but differ in one or more versions and have a newer timestamp?

* "update" may be more complicated than just "get the entry that exists and change it", possibly a task that gets fired off to delete entries older than N updates ago?